### DISCUSSION TITLE
Datasource github-tags fails to fetch releases when organization uses IP address allow list

### DISCUSSION THREAD

**Discussion by nikolaik** [renovatebot/renovate#20344] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

When trying to keep the github action https://github.com/bridgecrewio/checkov-action up to date in our workflow with renovate we get the error below. The error seems to indicate that they are using the [IP Allow List feature](https://docs.github.com/en/enterprise-cloud@latest/organizations/keeping-your-organization-secure/managing-security-settings-for-your-organization/managing-allowed-ip-addresses-for-your-organization) and the GitHub API does not allow renovate bot to fetch the tags, even though the repo is public.

Doing a simple test with the `gh` CLI seems to work fine:

```console
$ gh api /repos/bridgecrewio/checkov-action/git/refs/tags
[
  {
    "ref": "refs/tags/v1",
    "node_id": "MDM6UmVmMjU2NzM4MDE2OnJlZnMvdGFncy92MQ==",
    "url": "https://api.github.com/repos/bridgecrewio/checkov-action/git/refs/tags/v1",
    "object": {
      "sha": "1837edef0f1ae44e1998ed6bbe8ee4ef8738dd1f",
      "type": "tag",
      "url": "https://api.github.com/repos/bridgecrewio/checkov-action/git/tags/1837edef0f1ae44e1998ed6bbe8ee4ef8738dd1f"
    }
  },
(...)
```

Is this a bug in the GitHub API? It seems weird that the renovate app can not read the already public tags data.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Datasource unknown error
{
  "datasource": "github-tags",
  "packageName": "bridgecrewio/checkov-action",
  "err": {
    "message": "Although you appear to have the correct authorization credentials, the `bridgecrewio` organization has an IP allow list enabled, and 34.210.66.22 is not permitted to access this resource.",
    "stack": "Error: Although you appear to have the correct authorization credentials, the `bridgecrewio` organization has an IP allow list enabled, and 34.210.66.22 is not permitted to access this resource.\n    at GithubGraphqlDatasourceFetcher.doRawQuery (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/github/graphql/datasource-fetcher.js:79:29)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async GithubGraphqlDatasourceFetcher.doShrinkableQuery (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/github/graphql/datasource-fetcher.js:130:26)\n    at async GithubGraphqlDatasourceFetcher.doPaginatedQuery (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/github/graphql/datasource-fetcher.js:161:33)\n    at async GithubGraphqlDatasourceFetcher.getItems (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/github/graphql/datasource-fetcher.js:197:21)\n    at async Function.query (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/github/graphql/datasource-fetcher.js:29:23)\n    at async queryTags (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/github/graphql/index.js:8:17)\n    at async GithubTagsDatasource.getReleases (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/github-tags/index.js:58:28)\n    at async getRegistryReleases (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/index.js:57:17)\n    at async fetchReleases (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/index.js:241:23)\n    at async getPkgReleases (/home/ubuntu/renovateapp/node_modules/renovate/dist/modules/datasource/index.js:295:34)\n    at async lookupUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/lookup/index.js:53:45)\n    at async fetchDepUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/fetch.js:52:25)\n    at async /home/ubuntu/renovateapp/node_modules/p-map/index.js:57:22"
  }
}
DEBUG: Failed to look up github-tags dependency bridgecrewio/checkov-action (bridgecrewio/checkov-action)(packageFile=".github/workflows/checkov.yaml", dependency="bridgecrewio/checkov-action")
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by rarkins** [comment]
This is a GitHub problem and has been raised before. Unfortunately nobody could get GitHub to understand and fix it. Does not seem like anything Renovate can fix

**Comment by nikolaik** [comment]
Thanks for explaining. I searched the repo and could not find any relevant issues. Is this only relevant for request that renovate performas against graphql API or does the REST API suffer the same issue you think? Is it something that only only affects GitHub apps?

**Comment by nikolaik** [comment]
What is the most appropriate channel to submit this issue to GitHub?

**Comment by rarkins** [comment]
If you are an enterprise customer, submit via your company/enterprise support account.

In short - as I'm pretty sure you understand already - GitHub seems to have a bug when:
- An org has IP allow lists enabled, and
- The org has a public repo, and
- An app/bot which does not have an IP allow list attempts to GET the list of tags/releases for the repo

I will also try submitting this on behalf of Renovate's parent company's enterprise support account too. We may need to try multiple times until we get a first/second line support person at GitHub who can understand what we're trying to describe.

**Comment by rarkins** [comment]
FYI I have filed Ticket #2008128 with the following content:

The `bridgecrewio` org on GitHub has enabled IP restrictions.

https://github.com/bridgecrewio/checkov-action is a public repo within that org.

Renovate Bot (a GitHub App) does not have any IP list. When it tries to fetch tags/releases for that _public_ repo using authenticated GraphQL, GitHub.com's API returns this error:

```
  "err": {
    "message": "Although you appear to have the correct authorization credentials, the `bridgecrewio` organization has an IP allow list enabled, and 34.210.66.22 is not permitted to access this resource.",
  }
```

This seems to be in error. The repo is *public*, and I can fetch these same APIs if I have a non-app (e.g. personal user) account. A GraphQL query on this public repo should not be rejected due to IP allow lists.

**Comment by rarkins** [comment]
As expected, GitHub support unfortunately don't seem to understand the problem. Their response:

> The results you're seeing are expected. Organization members, including apps, can only access organization resources from IP addresses on the allow list, regardless of whether those resources are private or public. It is similar to how organization members in a SAML-enforcing organization can't access public repositories from personal devices if they can't authenticate via SAML from those devices.
> 
> This is because organization members may have additional access. We have been working on options to increase the level of access to public repositories available to organization members when they're authenticated to GitHub but not to the organization, in response to similar customer feedback. However, I can't promise what (if any) additional level of access this will result in or when this work will be completed. In all decisions, we are choosing to prioritize security and performance.
> 
> So, for this query to be successful, the IP address 34.210.66.22 must be permitted prior to the request being made. Please let me know if you have any questions, and I'll be more than happy to address them.

I have replied pointing out that the app is not a member of `bridgecrewio`, and even it was, such restrictions shouldn't apply to _other_ orgs which install the same app.

**Reply by rarkins** [reply, marked as answer]
GitHub's follow-up is to suggest that Bridgecrew should put their open source into a separate org just for their open source :(

**Comment by viceice** [comment]
i think we should move this to a discussion as it's not a renovate bug and we can't really do anything to bypass those restrictions.

**Reply by rarkins** [reply]
Converted

**Comment by stimmerman** [comment]
Did you ever find a solution to get this working @nikolaik?

I'm thinking along the lines of running renovate on a forked repository in my personal account en then use some Github action to mirror the dependency PRs to my upstream repo. 🤷🏻‍♂️

**Reply by nikolaik** [reply]
Sorry for not replying sooner. For the specific action https://github.com/bridgecrewio/checkov-action , we were able to switch us use the docker image directly in the Github Actions workflow, since the action is simply a small wrapper around it.
```yaml
# workflow.yaml
# ...
    job-name:
        # ...
        uses: docker://ghcr.io/bridgecrewio/checkov:2.5.19

**Reply by nikolaik** [reply]
The mirrored repo approach would work, but it would probably require some maintenance of build artifacts and/or tagging.
